### DISCUSSION TITLE
메모리계층, 스와핑, 페이지폴트

### DISCUSSION THREAD

**Discussion by jindream6128** [24tngus/CS_STUDY#44] [Operating System]
# 메모리 계층

- 레지스터 : CPU 내의 작은 메모리 (휘발성)
- 캐시 : CPU 내의 L1, L2, L3 캐시를 지칭 (휘발성)
- `주 기억 장치` : RAM (휘발성)
- `보조 기억 장치` : HDD, SDD (비휘발성)

![ㅇ1](https://github.com/24tngus/CS_STUDY/assets/101094583/8733951e-d7fc-4db8-9d03-1f825a4d1220)

계층이 존재하는 이유

1. 더 빠른 접근과 처리속도가 증가
    1. 자주 사용하는 것을 많이 쓰기 대문에, 좀 더 작은 캐시메모리에 해당 데이터가 있다면 빠르게 해당 데이터에 접근이 가능해지며 빨라진다.
2. 비용의 효율성
    1. 적절한 배치로 가성비
3. 자원의 효율적 사용
    1. 자주 접근하는 데이터는 빠른 메모리에 저장하며 자원을 효율적 사용
    2. 잘 접근하지 않는 데이터는 비교적 느린 하드디스크에 저장하여 자원을 낭비하지 않는다
    

# 가상메모리와 스와핑, 페이지폴트 그리고 스레싱

가상 메모리 ?

→ OS에서 사용되는 메모리 관리 기법 중 하나이다.

→ 실제로 이용 가능한 메모리 주소를 추상화 하여 이를 사용하는 사용자들에게는 매우 큰 메모리 처럼 보이게 만드는 것.

→ MMU와 페이지 테이블에 의해 실제 주소로 바뀌게 된다

페이지 : 가상 메모리를 사용하는 최소 크기 단위

프레임 : 실제 디스크나, 메모리를 사용하는 최소 크기 단위

![ㅇ2](https://github.com/24tngus/CS_STUDY/assets/101094583/3148c393-0d37-4c51-bbd4-2b07baf1478a)

페이지 테이블 

→ 가상 메모리는 가상 주소, 실제 주소가 매핑되어 있는 페이지 테이블로 관리, 이때 속도 향상을 위해 캐싱 계층인 `TLB(Translation Lookaside Buffer) 변환 색인 버퍼`를 사용한다.

→ 일종의 캐시역할 

![ㅇ3](https://github.com/24tngus/CS_STUDY/assets/101094583/9229a9d7-950f-463f-9d77-288f22aa40a0)

페이지 폴트 

→ 가상 메모리는 작은 메모리 → 매우 큰 메모리처럼 보이게끔 한다.

→ 실제로 참조하려는 메모리 영역이 없을 수도 있다 !

이때 페이지 폴트가 발생한다 !

스와핑

→ 페이지 폴트가 발생했을 때 당장 사용하지 않는 영역을 하드디스크로 옮기고 `하드디스크의 일부분을 마치 메모리`처럼 불러와 쓰는 것.

![ㅇ4](https://github.com/24tngus/CS_STUDY/assets/101094583/3e0ea5a6-e9f2-4aa9-ad9d-97967dc63080)

페이지 폴트의 과정

![ㅇ5](https://github.com/24tngus/CS_STUDY/assets/101094583/5313c32e-4377-4fe1-97c2-bc10daa6c9ff)

1. 가상 주소에 접근 하였으나, 해당 페이지가 없다면, Trap(소프트웨어 인터럽트) 이 발생 이를 OS에 알린다.
2. OS는 실제 디스크로부터 사용하지 않는 프레임을 찾고
3. 해당 프레임을 실제 메모리에 가져와서 특정 페이지와 교체한다 (스와핑 발생)
4. 페이지 테이블을 리셋하고, 다시 명령어를 시작한다.

스레싱

- 스레싱은 메모리의 페이지 폴트율이 높은 것을 의미
- 스레싱은 /메모리에 너무 많은 프로세스가 동시에 올라가면/ 페이지 폴트가 많이 발생/ 그에 따른 스와핑 과정이 많이 일어나며/ 발생하게 된다.
- 이렇게 되면 /CPU 이용률은 낮아지고 /OS는 CPU를 더 쓰기 위해 더 많은 프로세스를 올리는데/ 이게 반복되면 스레싱이 발생한다.

해결과정 

1. HW 변경
2. 작업세트
    1. 프로세스의 과거 사용이력을 기반으로 많이 사용하는 페이지 집합을 만들어 한꺼번에 미리 메모리에로 전송
3. PFF
    1. 상한선과 하한선을 만들어 페이지 폴트의 빈도 조절
    
![ㅇ6](https://github.com/24tngus/CS_STUDY/assets/101094583/49b27979-529e-42c9-b506-63acc5ebec2a)
    
    왜? 가성메모리가 필요한가
    
    1. 주기억장치의 효율적인 관리 (스와핑) → RAM
    - 하드디스크를 주기억장체에 대한 캐시로 설정하여, 필요한 데이터만 빠르게 가지고와 주 기억장치를 효율적으로 관리
    
    1. 메모리 관리의 단순화
    - 각 프로세스마다 가상메모리의 통일된 주소 공간을 배정
    
    1. 메모리 용량 및 안정성 보장
    - 한정된 공간의 램이 아닌 거의 무한한 가상 메모리 공간을 배정

**Comment by berryberries** [comment]
확인했습니다!
